### PR TITLE
feat: support autoconversion of Entity to Key for purposes of delete & delete_multi

### LINKED ISSUE #122: Allow passing Entity types to delete_multi in addition to Key types.
Branched off from https://github.com/googleapis/python-datastore/issues/118

`delete` and `delete_multi` require a List of `Key`, but could possibly take `Entity`

Currently:
```
query = client.query(kind='Task')
query.keys_only()
entities = query.fetch()
keys = [i.key for i in entities]
client.delete_multi(keys)
```

Proposed:
```
query = client.query(kind='Task')
query.keys_only()
entities = query.fetch()
client.delete_multi(entities)
```

This would also better support conversion from NDB as fetch can return entities or keys based on the keys_only flag.

### PR DESCRIPTION
**Pull request by crwilcox**
Fixes #122

Can discuss if we should
- have a warning
or
- modify documented type of delete and delete_multi to mention they can take `Entity`

```
❯ python tests/system/test_delete.py
Deleting 20 entities: 5121006972698624, 5150910414061568, 5629847317512192, 5635426882682880, 5644748102565888, 5660593176444928, 5670520389369856, 5680769993277440, 5693029306335232, 5697954694299648, 5701878616686592, 5708050082037760, 5709769612460032, 5744781917421568, 5746397093560320, 5746853299617792, 5754048175144960, 5756417252261888, 5762132310228992, 6251525647106048
/Users/crwilcox/workspace/python-datastore/google/cloud/datastore/client.py:682: UserWarning: A list of :class:`google.cloud.datastore.key.Key` is expected. Extracting Key from :class:`google.cloud.datastore.entity.Entity`.
  warnings.warn(message, UserWarning)
```

### CODE REVIEW

File: google/cloud/datastore/client.py [hunk 1]
```diff
@@ -671,6 +671,16 @@ def delete_multi(self, keys, retry=None, timeout=None):
...
+            if type(key) is Entity:
+                # While the interface is technically to take an iterable of keys
+                # if an entity is provided the key can be extracted.
+                message = (
+                    "A list of :class:`google.cloud.datastore.key.Key` is expected. "
+                    + "Extracting Key from "
+                    + ":class:`google.cloud.datastore.entity.Entity`."
+                )
+                warnings.warn(message, UserWarning)
+                key = key.key
```
Comment (wescpy) [inline google/cloud/datastore/client.py:683]: - [L683] This works but may result in occasional snarky comments. I would *not*, however, rewrite it as `key = getattr(key, 'key')` cuz that's worse.
- [L646] Would any changes be req'd in this part of the docstring?

**Review by wescpy — APPROVED** [review]
LGTM

File: google/cloud/datastore/client.py (hunk 1 above)
Comment (crwilcox) [reply google/cloud/datastore/client.py:683]: >  may result in occasional snarky comments.

@wescpy I don't quite understand. Could you elaborate.

As for docstr, that is part of the open conversation. We can either

1. make this a stealth change to help smooth migration. As written with the warning, I wouldn't want to add to docs
1. say this is actually supported, in which case, no warning should happen, but we should specify in types you can hand an entity as well.

I would be fine to call this supported, but started from the 'smoothing' over as lightest touch.

File: google/cloud/datastore/client.py [hunk 2]
```diff
@@ -671,6 +671,16 @@ def delete_multi(self, keys, retry=None, timeout=None):
             current.begin()
 
         for key in keys:
+            if type(key) is Entity:
```
Comment (tseaver) [inline google/cloud/datastore/client.py:674]:
```suggestion
            if isinstance(key, Entity):
```

**Review by tseaver — DISMISSED** [review]
Needs unit tests for this feature.

**Comment by crwilcox** [comment]
@tseaver sort of waiting until we agree if to go with full support or sneaky support. Do you have an opinion? I am not terrified about adding type coercion myself, and this will make migration from NDB a bit smoother.

File: google/cloud/datastore/client.py (hunk 1 above)
Comment (wescpy) [reply google/cloud/datastore/client.py:683]: Snarky meaning some people may criticize naming choice in `key = key.key` while it's perfectly valid IMO. Anyway, it was an aside and a mostly irrelevant comment... pls ignore.

**Comment by tseaver** [comment]
@crwilcox I would actually go with full support (no warning, update the docstring).

**Review by dmahugh — APPROVED** [review]
LGTM

**Review by engelke — APPROVED** [review]
LGTM

**Pull request merged**